### DISCUSSION TITLE
stop rules from "self triggering" of different trigger

### DISCUSSION THREAD

**Discussion by GsakuL** [arendst/Tasmota#16607] [Q&A]
Hello
I have searched GH Discussions, Issues, Discord and Docs, but I have not found any hints/solutions fpr my problem yet.

Essentially, I want to add some way to have my iFan03 only run for X ammount of time. Currently I'm trying this with an Event and Timer2. But the problems lies within Rule2: when I trigger the event, the Timer2 gets changed correctly and activated. But it triggers on FanSpeed#Data for the incomming change, and instantly resets the Timer2. (i want a "mannual override")
I could filter that Event on Data!=1, but that seems unclean. 
The [Docs](https://tasmota.github.io/docs/Rules/#rule-syntax) also says "`BREAK` will stop the execution of the remaining rules that follow this rule within the rule set".
Does that mean, it doesnt make a difference, because it's a new execution?
Might a second posibility be, to put the offending Data-Event in a seperate rule, and temporarily disabling it during my event? - But that also seems rather inconvinient, espacially, if i have to consider rule placement there.
I also guess that RuleTimer* has the same symptoms, as Timer*.
(I have a custom compiled Image with Rules, and IF)
```json
{"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":335,"Free":176,"Rules":"ON Power1#Boot Backlog var1 0; var2 0 ENDON ON Event#SlowDown DO IF (var1>1) FanSpeed 1 ENDIF ENDON ON Clock#Timer=1 DO Event SlowDown ENDON ON Power1#State DO var2 %value% ENDON ON Event#CT DO Backlog0 IF (var2>0) Power1 0; Delay 500 ENDIF; Power1 1; Delay 150; Power1 0; Delay 150; Power1 1 ENDON ON Clock#Timer=2 DO FanSpeed 0 ENDON"}}
```
```js
Rule1
  ON Power1#Boot Backlog var1 0; var2 0 ENDON
  ON Event#SlowDown DO IF (var1>1) FanSpeed 1 ENDIF ENDON
  ON Clock#Timer=1 DO Event SlowDown ENDON
  ON Power1#State DO var2 %value% ENDON
  ON Event#CT DO Backlog0 IF (var2>0) Power1 0; Delay 500 ENDIF; Power1 1; Delay 150; Power1 0; Delay 150; Power1 1 ENDON
  ON Clock#Timer=2 DO FanSpeed 0 ENDON
```

```json
{"Rule2":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":149,"Free":362,"Rules":"ON Event#Until DO Backlog0 FanSpeed 1; Timer2 {\"Enable\":1,\"Time\":\"%value%\"} break ON FanSpeed#Data DO Backlog Timer2 {\"Enable\":0}; var1 %value% ENDON"}}
```
```js
Rule2
  ON Event#Until DO Backlog0 FanSpeed 1; Timer2 {"Enable":1,"Time":"%value%"} break
  ON FanSpeed#Data DO Backlog Timer2 {"Enable":0}; var1 %value% ENDON
```

Edit (my config, for features. Yes, i know..):
```c
#ifndef USE_RULES
#define USE_RULES
#endif

#ifndef USE_EXPRESSION
#define USE_EXPRESSION
#endif

#ifndef SUPPORT_IF_STATEMENT
#define SUPPORT_IF_STATEMENT
#endif

#ifndef USE_SONOFF_IFAN
#define USE_SONOFF_IFAN
#endif
```

**Comment by sfromis** [comment]
Using `break` is only useful when multiple triggers act on the same data message, like multiple triggers on same field, or triggers on closely related fields, like from the same sensor message. But other than this, the rules are completely independent, and the whole event handling starts from zero with every message with the potential of triggering something. Rules are no sequence, each is unrelated to whatever other rules happens to be nearby, with a few special cases like break or one-shot.

It is not clear to me why you even use FanSpeed#Data which I's see as secondary message "for historical purposes", my starting point would instead be to take control over the activation of the fan, and thus decide when to do it or not.

**Reply by GsakuL** [reply]
oh. ok. that makes sense. The Docs only showed such examples, but it did not explain it in such detail.

---

I initally used FanSpeed#Data to track if the Speed would be above 1. I Use this, because the fan is "right nexto to" a window, and in summer, that window might be open, and Speed 2 and especially 3 are quite nosy. Where i live, beedroom windows are next to each other, so neighbours possibly hear it. So i want to turn down the speed at a certain time (Timer1 on repeat), if it's too high. I see an alternative that doesn't use IF, with Data-comparison at triger side with multiple Events and enable/disable Timer1.

Thinking of that, Timers are persistent, so changing them often might not be a "good idea" and "wear out" storage?
So RuleTimer might be better anyway for my "Until" Event. but i have not found a Limit on the countdown size/time, is there any? like int32?

But what do mean by `take control over the activation`? Can you hook between Commands? And "cancel/abort" them? And if so, where can read about it? Because then I have to think about "optimizing" my rules more.

**Reply by sfromis** [reply]
What I mean with "taking control" is to redirect the triggering (like a button) to a rule, which then can decide which action to take. Note that this is meant in general, I don't know the details of fan handling.

Yeah, timers are persistent, and contribute to flash wear, but so does Power, unless you disable saving of that to flash.

Note: Using Delay in rules is not reliable for timing, the interval may be cut short.

**Reply by GsakuL** [reply]
Ah yes. Of course. But I have the Fan connected via HomeAssistent and the Tasmota integration. Replacing the Fan command with a custom event would have it's own effort (manual MQTT Device in config files, but yeah).

I do not care for saving that, since it's default off/0 anyways on reboot. Good point.

Yes, I know. I just put that in lazily. I know fast switching relays may also be bad. But i have added some leyway. And I use this maybe once a month during summer. But now, winter coming up, more often. This is just there, to have the main Light be switched to warm-white, instead of mixed/off-white. Sometimes more pleasant.
I'll look to optimize this.
